File: apps/kira-studio/internal/bridge/events.go
package bridge

import (
	"github.com/kirathecat/kira-studio/apps/kira-studio/internal/appcore"
	"github.com/kirathecat/kira-studio/apps/kira-studio/internal/dbmcp"
	"github.com/kirathecat/kira-studio/apps/kira-studio/internal/storage/model"
	"github.com/kirathecat/kira-studio/internal/appevent"
	"github.com/kirathecat/kira-studio/internal/metrics"
)

// Channel holds today's exact IPC channel strings (packages/shared/protocol/ipc.ts's IPC const), which
// are the Wails event names verbatim (P52 §7.1) — the renderer's subscribe mechanism changes, the
// wire name does not. The ones re-exported below (P103 Part 3, widened P116) are byte-identical
// strings shared with Kira Space, now defined once in repo-root internal/appevent — re-exported
// under their own names here so no call site anywhere in this package has to change.
const (
	ChannelOpenSettings           = appevent.ChannelOpenSettings
	ChannelNewConnection          = "kira:menu:new-connection"
	ChannelNewRequest             = "kira:menu:new-request"
	ChannelImportPostman          = "kira:menu:import-postman"
	ChannelImportDataGrip         = "kira:menu:import-datagrip"
	ChannelToggleProjectPanel     = appevent.ChannelToggleProjectPanel
	ChannelToggleOperationsPanel  = "kira:menu:toggle-operations-panel"
	ChannelCommandPalette         = "kira:menu:command-palette"
	ChannelTabNext                = appevent.ChannelTabNext
	ChannelTabPrev                = appevent.ChannelTabPrev
	ChannelTabClose               = appevent.ChannelTabClose
	ChannelViewFind               = "kira:menu:view-find"
	ChannelViewRefresh            = "kira:menu:view-refresh"
	ChannelViewRun                = "kira:menu:view-run"
	ChannelViewRunAll             = "kira:menu:view-run-all"
	ChannelViewFormat             = "kira:menu:view-format"
	ChannelFlushBeforeClose       = appevent.ChannelFlushBeforeClose
	ChannelWindowFlushBeforeClose = appevent.ChannelWindowFlushBeforeClose
	ChannelConnectionState        = "kira:connection:state"
	ChannelMetadataInvalidated    = "kira:connection:metadataInvalidated"
	ChannelConnectionsChanged     = "kira:connections:changed"
	ChannelSettingsChanged        = appevent.ChannelSettingsChanged
	ChannelLayoutChanged          = appevent.ChannelLayoutChanged
	ChannelOpUpdate               = "kira:op:update"
	ChannelAppMetrics             = appevent.ChannelAppMetrics
	ChannelSchemaChanged          = "kira:schema:changed"
	// ChannelGrpcCall is P11 D8's own new push channel — a server-streaming call's coalesced
	// message batches, delivered with EmitTo (one window only, GrpcService.Call's own emitter
	// call) rather than through this file's Events/Sources machinery: it has exactly one producer
	// (the RunOp closure calling grpcclient.ServerStream), not a startup-wired long-lived source,
	// so it needs no Sources entry and no Attach subscription.
	ChannelGrpcCall = "kira:grpc:call"
	// ChannelDbMcpApproval is M2 §7.1's own push channel — the prompt-mode approval queue's live
	// snapshot: one FIFO of pending approvals, one presented at a time, each with its own timeout.
	ChannelDbMcpApproval = "kira:dbmcp:approval"
	// ChannelTerminal is P83's own push channel — one terminal's output and its exit, EmitTo'd to
	// the one window that opened it, flushed on the same 60ms/256-batch/terminal-event shape as
	// ChannelGrpcCall above (internal/bridge/terminal.go's own coalescer).
	ChannelTerminal = appevent.ChannelTerminal
	// ChannelCustomScriptsChanged is P85's own list-changed broadcast — CustomScriptsService's own
	// Create/Update/Remove Emit (not EmitTo) the full list, ChannelConnectionsChanged's own shape,
	// so every window's tab-strip dropdown stays in sync with a script added or removed elsewhere.
	ChannelCustomScriptsChanged = "kira:customScripts:changed"
	// ChannelMaskRulesChanged is P108 Part 12 F18's own push channel — MaskRulesService's own
	// Upsert/Remove/RegenerateKey Emit (not EmitTo) one connection's own rule set, ChannelSchemaChanged's
	// own per-connection shape rather than ChannelCustomScriptsChanged's flat app-wide list, so a
	// second window's Privacy tab / grid header menu / grid preview stay in sync with a rule changed
	// elsewhere.
	ChannelMaskRulesChanged = "kira:maskRules:changed"
	// ChannelAgentSessions is P86 §11's own app-wide broadcast — every live Claude Code session
	// across every window, Emit'd (not EmitTo) whenever terminal.Registry.OnChange fires, so the
	// status-bar widget in every window agrees on the same count.
	ChannelAgentSessions = "kira:agent:sessions"
	// ChannelAgentEvent is P86 §8.4's own per-hook broadcast — one Claude Code hook firing for one
	// tab, Emit'd (not EmitTo, unlike ChannelTerminal) since AgentHooksService.onEvent has no
	// window to address: a hook event is filtered by the receiving window against terminals it
	// owns instead (state/agentSessions.ts's own reducer, keyed by terminalId).
	ChannelAgentEvent = "kira:agent:event"
	// ChannelKeepAwake is P87 §3.2's own app-wide broadcast — the titlebar toggle's state, Emit'd
	// (not EmitTo) exactly like ChannelAgentSessions: one machine, one assertion, so every window's
	// titlebar button must agree. Process-scoped, never persisted: an OS power assertion that
	// outlives the reason a user made it is a surprise, and the persistent half of this feature is
	// the Settings toggle. P116: byte-identical to Kira Space's own channel of the same name,
	// hoisted to repo-root internal/appevent.
	ChannelKeepAwake = appevent.ChannelKeepAwake
	// ChannelApiDataChanged is P112's own broadcast — every API-client mutation (collections,
	// saved requests, variables, environments) Emits it with the scopes that mutation touched, so
	// every window's TanStack Query cache invalidates exactly those keys (apidata.go).
	ChannelApiDataChanged = "kira:api:dataChanged"
)

// ChannelEngineState is declared for completeness and deliberately never emitted: nothing in
// src/main sends it and nothing in apps/kira-studio/frontend/src subscribes to it (P56 D5). P57 deletes it.
const ChannelEngineState = "kira:engine:state"

// Sources are the five push producers P55 left as seams (P55 D15). Each is a one-method
// interface, small enough to drive with a plain recorder rather than a real service — there is
// no dedicated events_test.go in this package (P56's own test-bar pruning removed it; this
// comment used to still claim otherwise, corrected here per P8 C8).
type Sources struct {
	Connections interface {
		OnStateChange(func(model.ConnectionState)) func()
		OnMetadataInvalidated(func(string)) func()
		OnListChanged(func([]model.ConnectionSummary)) func()
	}
	Oplog interface {
		OnUpdate(func(model.OpRecord)) func()
	}
	Metrics interface {
		OnSample(func(metrics.Sample)) func()
	}
	DbMcp interface {
		OnApprovalChange(func(dbmcp.ApprovalSnapshot)) func()
	}
}

// Events is the Go analogue of the five push-side listeners src/main/index.ts wires (:58, :99-
// 102) plus the menu/quit signal channels. It is the only thing in this package that emits.
// Signal/SignalTo/Broadcast come from the embedded *appevent.Events core (P103 Part 3); emit is
// kept alongside it for Attach's five producer subscriptions and SettingsChanged below, neither of
// which the shared core exposes a raw Emit for.
type Events struct {
	*appevent.Events
	emit appcore.Emitter
}

func NewEvents(e appcore.Emitter) *Events {
	return &Events{Events: appevent.NewEvents(e), emit: e}
}

// Attach subscribes to every producer in s and returns one detach that unsubscribes all of them.
// It is called once at startup and detached first in the quit teardown (P56 D3), so nothing
// emits into a half-torn-down app.
func (ev *Events) Attach(s Sources) (detach func()) {
	unsubState := s.Connections.OnStateChange(func(st model.ConnectionState) {
		ev.emit.Emit(ChannelConnectionState, st)
	})
	unsubInvalidated := s.Connections.OnMetadataInvalidated(func(connectionID string) {
		ev.emit.Emit(ChannelMetadataInvalidated, connectionID)
	})
	unsubList := s.Connections.OnListChanged(func(list []model.ConnectionSummary) {
		ev.emit.Emit(ChannelConnectionsChanged, list)
	})
	unsubOplog := s.Oplog.OnUpdate(func(rec model.OpRecord) {
		ev.emit.Emit(ChannelOpUpdate, rec)
	})
	unsubMetrics := s.Metrics.OnSample(func(sample metrics.Sample) {
		ev.emit.Emit(ChannelAppMetrics, sample)
	})
	unsubDbMcpApproval := s.DbMcp.OnApprovalChange(func(snap dbmcp.ApprovalSnapshot) {
		ev.emit.Emit(ChannelDbMcpApproval, toWireApprovalSnapshot(snap))
	})

	return func() {
		unsubState()
		unsubInvalidated()
		unsubList()
		unsubOplog()
		unsubMetrics()
		unsubDbMcpApproval()
	}
}

// Signal/SignalTo/Broadcast (D6: nil payloads, not {}) come from the embedded *appevent.Events
// core — Electron's own sendToFocusedWindow(channel) (18fe7bb^:src/main/menu.ts:5-8; preload's
// onSignal discards arguments, src/preload/index.ts:35-39). The menu's twelve signal channels are
// Signal's only caller (P8 C9): a background window no longer reacts to a command aimed at
// whichever window the user was actually looking at — Cmd+W closing a tab in every open window, or
// Cmd+Return running a console statement in a window the user never touched, were the concrete
// symptoms (F2). This is the split C9 exists for: Broadcast (the quit handshake's own trigger) and
// the six state-change broadcasts in Attach/SettingsChanged/LayoutService.Set all stay on Emit —
// every window genuinely needs those, unlike a menu command.

// SettingsChanged broadcasts the merged settings unconditionally — SettingsService.Set's own
// job, factored out here so its own file stays a thin wrapper.
func (ev *Events) SettingsChanged(s model.Settings) {
	ev.emit.Emit(ChannelSettingsChanged, s)
}

File: packages/shared/protocol/events.ts
import type { MaskRule } from '../domain/mask';
import type { VariableScope } from '../domain/variables';

/** The Go→renderer push channels (`apps/kira-studio/internal/bridge/events.go`'s constants, verbatim).
 *  Formerly the push half of ipc.ts's IPC const; the request/response half retired with
 *  window.kira, and the wire types for bound calls now come from the generated bindings (P57 D7).
 *  `kira:engine:state` and `kira:port` are absent — both are dead channels retired by D6/D3. */
export const CHANNEL = {
  openSettings: 'kira:open-settings',
  newConnection: 'kira:menu:new-connection',
  newRequest: 'kira:menu:new-request',
  importPostman: 'kira:menu:import-postman',
  importDataGrip: 'kira:menu:import-datagrip',
  toggleProjectPanel: 'kira:menu:toggle-project-panel',
  toggleOperationsPanel: 'kira:menu:toggle-operations-panel',
  commandPalette: 'kira:menu:command-palette',
  // C9 D5: ⌘P's fuzzy file finder — the View menu item above Command Palette.
  quickOpen: 'kira:menu:quick-open',
  tabNext: 'kira:menu:tab-next',
  tabPrev: 'kira:menu:tab-prev',
  tabClose: 'kira:menu:tab-close',
  viewFind: 'kira:menu:view-find',
  viewRefresh: 'kira:menu:view-refresh',
  viewRun: 'kira:menu:view-run',
  viewRunAll: 'kira:menu:view-run-all',
  viewFormat: 'kira:menu:view-format',
  appFlushBeforeClose: 'kira:app:flush-before-close',
  windowFlushBeforeClose: 'kira:window:flush-before-close',
  connectionState: 'kira:connection:state',
  connectionMetadataInvalidated: 'kira:connection:metadataInvalidated',
  connectionsChanged: 'kira:connections:changed',
  settingsChanged: 'kira:settings:changed',
  layoutChanged: 'kira:layout:changed',
  opUpdate: 'kira:op:update',
  appMetrics: 'kira:app:metrics',
  schemaChanged: 'kira:schema:changed',
  // P11 D8: a server-streaming call's coalesced message batches, delivered via EmitTo (one window
  // only) — the one genuinely new push channel this phase adds.
  grpcCall: 'kira:grpc:call',
  // G1 §3.6/D19: the pairing prompt's live queue snapshot and the Connected editors pane's list.
  gitPairing: 'kira:git:pairing',
  gitClientsChanged: 'kira:git:clients',
  // C7 D7: a repository-wide search's coalesced file groups, delivered via EmitTo (one window
  // only) — grpcCall's own shape, restated for a payload that shares no field with it.
  codeSearch: 'kira:code:search',
  // M2 §7.1: the prompt-mode approval queue's live snapshot — gitPairing's own shape.
  dbMcpApproval: 'kira:dbmcp:approval',
  // P83 §3.2: one terminal's coalesced output and its exit, delivered via EmitTo (one window
  // only) — codeSearch's own shape, restated for a byte payload.
  terminal: 'kira:terminal:data',
  // P85 §9.3: the custom-scripts list changed — connectionsChanged's own shape (Emit, not EmitTo,
  // so every window's tab-strip dropdown stays in sync).
  customScriptsChanged: 'kira:customScripts:changed',
  // P108 Part 12 F18: one connection's mask-rule set changed (Upsert/Remove/RegenerateKey) —
  // schemaChanged's own per-connection shape (Emit, not EmitTo), so a second window's Privacy tab,
  // grid header menu and grid preview all stay in sync rather than reading a stale rule list.
  maskRulesChanged: 'kira:maskRules:changed',
  // P86 §11: every live Claude Code session across every window, Emit'd (not EmitTo) whenever
  // terminal.Registry.OnChange fires — customScriptsChanged's own shape, app-wide by definition.
  agentSessions: 'kira:agent:sessions',
  // P86 §8.4: one Claude Code hook firing for one tab, Emit'd (not EmitTo) — a hook event has no
  // window to address, so every window's own store filters by terminalId against tabs it owns.
  agentEvent: 'kira:agent:event',
  // P87 §3.2: the titlebar keep-awake toggle's own state, Emit'd (not EmitTo) exactly like
  // agentSessions — one machine, one assertion, so every window's button must agree.
  keepAwake: 'kira:keepAwake:changed',
  // P112: every API-client mutation (collections, saved requests, variables, environments) Emits
  // this with the scopes it touched, so every window's TanStack Query cache invalidates exactly
  // those keys — customScriptsChanged's own shape, generalised to a batch of scopes per event.
  apiDataChanged: 'kira:api:dataChanged',
} as const;

/** Summed across every process metrics.Sample covers (P56's ticker) — a single app-wide readout
 *  for the status bar, not a per-process breakdown. No generated binding carries this shape: it
 *  is emitted (`kira:app:metrics`), never returned from a bound call, so it has no home in any
 *  service's models.ts and stays hand-written here instead (formerly ipc.ts's own type). Mirrors
 *  `apps/kira-studio/internal/metrics.Sample`'s JSON tags.
 *
 *  cpuPercent is normalized to the machine's whole capacity (0-100), not the per-core-sum macOS's
 *  own Activity Monitor shows in its per-process "% CPU" column (which reads up to `logicalCPUs`
 *  times higher for the same load) — StatusBar.vue's tooltip states this explicitly since it's the
 *  exact cross-check a user is likely to make. memoryBytes is RSS everywhere except darwin, where
 *  it is phys_footprint — Activity Monitor's own "Memory" column, not its "Real Memory" one (P7
 *  F1/D2). logicalCPUs and processCount exist so the tooltip can say what cpuPercent is a
 *  percentage *of* and how many processes memoryBytes covers (P7 F6/D6). */
export interface AppMetricsSample {
  cpuPercent: number;
  memoryBytes: number;
  logicalCPUs: number;
  processCount: number;
}

// P112: `kira:api:dataChanged`'s own payload (internal/bridge/apidata.go's ApiDataChange/
// ApiDataChanged, field for field). Never a bound-call arg or return type, so — AppMetricsSample's
// own comment applies verbatim — there is no generated binding to reuse.
export type ApiDataChange =
  | { kind: 'tree' }
  | { kind: 'savedRequest'; itemId: string }
  | { kind: 'variables'; scope: VariableScope; ownerId: string }
  | { kind: 'environments' };

export interface ApiDataChangedEvent {
  changes: ApiDataChange[];
}

/** `kira:terminal:data`'s own payload (internal/bridge/terminal.go's TerminalEvent, field for
 *  field) — one coalesced chunk of a session's output, or its exit. Never a bound-call arg or
 *  return type, so it has no generated binding — a plain interface here, mirroring
 *  AppMetricsSample just above: this event crosses no storage boundary and is never restored, so
 *  there is nothing for a zod schema to guard that this interface doesn't already state. */
export interface TerminalEvent {
  terminalId: string;
  data?: string; // base64, absent on the exit event
  exited: boolean;
  exitCode?: number;
  error?: string;
}

/** `kira:maskRules:changed`'s own payload (internal/bridge/maskrules.go's MaskRulesChangedEvent,
 *  field for field) — one connection's own rule set, sent whole on every Upsert/Remove/
 *  RegenerateKey. `keyRegenerated` is set only by RegenerateKey — the one case
 *  state/maskRules.ts's own applyRemote must also drop that connection's cached correlation key. */
export interface MaskRulesChangedEvent {
  connectionId: string;
  rules: MaskRule[];
  keyRegenerated: boolean;
}
